### PR TITLE
Allow larger datastore dumps

### PR DESCRIPTION
**Pull request by wardi**
This is a small improvement to datastore that allows large CSV dumps and shrinks the maximum memory used by paging over the data and streaming the output CSV

### CODE REVIEW

**Comment by mattfullerton** [comment]
:+1: Thanks for this!

**Comment by amercader** [comment]
In the original code it seems like people could define the limit on the request (eg I just want the first 10 rows). If I'm understanding your changes correctly people now will always get all the rows regardless of `limit` as it is used as the page size. So passing the `limit` parameter would just affect the internal paging size. Shouldn't we separate `limit` and `page`?

**Comment by wardi** [comment]
You're right. Are those parameters documented? (Do we *want* to keep them?) The way they were written made it easy to consume all the memory on the web server if a user requests enough rows.

I can add them back but I would like to document them at least.

**Comment by amercader** [comment]
There should probably be a hard limit either way, with params or without, even if it's a big one (1m ?)
Your offset is not actually used on the following loops if there is an `offset` passed as param. I fine dropping support for the params to simplify things.
What would probably be more useful for selective dumps is allow an option to download as CSV on `datastore_search`.

**Comment by wardi** [comment]
I agree, a `datastore_search` CSV output option would be more flexible. It doesn't fit the action API protocol though. Ideally we could get any of our (paginated or not) API calls' data returned as a CSV streamed back to the end user.

Let me just add those options back here in case someone is using them.

On another topic: Should I be including a UTF-8 BOM here so that Excel can open these files? It would be a user-visible change. Add an option for that at the same time?

**Comment by amercader** [comment]
I personally would not add it by default. I'm happy with an option to enable it optionally.

**Comment by wardi** [comment]
There's no reason for a hard limit on this call. If the client keeps reading the data this code will continue looping to the next 10k records until the end. There's no benefit in making the client send multiple requests.

**Comment by amercader** [comment]
Fair enough, we'll assume there are no memory leaks there :)

**Comment by TkTech** [comment]
@wardi regarding incompatibility with the existing api, we could support `Accept: text/csv` with a utility wrapper around certain actions. A quick `LazyCSVWriter` that doesn't write the headers until the first dict/row is consumed to get the structure? Makes the assumption that every row will be identical.

File: ckanext/datastore/controller.py [hunk 1]
```diff
@@ -20,27 +23,34 @@ def dump(self, resource_id):
...
+            }
+
+            action = p.toolkit.get_action('datastore_search')
+            try:
+                result = action(context, data_dict)
+            except p.toolkit.ObjectNotFound:
+                base.abort(404, p.toolkit._('DataStore resource not found'))
+
+            if not wr:
+                pylons.response.headers['Content-Type'] = 'text/csv'
```
Comment (amercader) [inline ckanext/datastore/controller.py]: `Content-Type: text/csv; charset=utf-8` ?

**Pull request merged**